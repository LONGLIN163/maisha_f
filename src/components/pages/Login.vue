<template>
    <div>
       <van-nav-bar
        title="Login"
        left-text="back"
        left-arrow
        @click-left="goBack"
        />

        <div class="register-panel">
        <van-field
            v-model="username"
            label="User Name"
            icon="clear"
            placeholder="Please input user name"
            :error-message="usernameErrorMsg"
            required
            @click-icon="username = ''"
        />

        <van-field
            v-model="password"
            type="password"
            label="Password"
            placeholder="Please input password"
            :error-message="passwordErrorMsg"
            required
        />
        <div class="register-button">
            <van-button 
            type="primary" 
            @click="loginAction" 
            size="large" 
            :loading="openLoading" 
            loading-text="loading"
            >Login</van-button>
        </div>
       </div>

    </div>
</template>

<script>
    import axios from 'axios'
    import serviceApi from '@/serviceApi.config.js'
    import {Toast} from 'vant'

    export default {
        data() {
            return {
                username: '',
                password: '',
                openLoading:false, // controller register loading status
                loadingText:'',
                usernameErrorMsg:'', // name validation
                passwordErrorMsg:'',  // password validation
            }
        },
        created(){
            if(localStorage.userInfo){
                    Toast.success('You r logged in!')
                    this.$router.push('/')
            }
        },
        methods: {
            goBack() {
                this.$router.go(-1)   
            },
            checkForm(){
                let isOk= true
                if(this.username.length<5){
                    this.usernameErrorMsg="username can not be less than 5 digit"
                    isOk= false
                }else{
                    this.usernameErrorMsg=''
                }
                if(this.password.length<6){
                    this.passwordErrorMsg="username can not be less than 6 digit"
                    isOk= false
                }else{
                    this.passwordErrorMsg=''
                }
                return isOk
            },
            loginAction(){
               this.checkForm() && this.login() // only register when no problem
            },
            login(){
                this.openLoading=true // start loading, prevent resubmit 2
                axios({
                    url:serviceApi.userLogin,
                    method:'POST',
                    data:{
                       userName:this.username,
                       password:this.password,
                    }
                })
                .then(res => {
                    if(res.data.code==200 && res.data.message){
                        // we need a promise to wait the login info get stored, then do the other thing
                        new Promise((resolve)=>{
                        // use localStorage to store loged in info
                        localStorage.userInfo={userName:this.username}
                        setTimeout(()=>{resolve()},500) 
                        }).then(()=>{ // if saved success,then do the other things
                            Toast.success('Login success!')
                            this.$router.push('/') // prevent resubmit 1
                        }).catch(err=>{
                            Toast.fail('Login status saved failed')
                            console.log(err)
                        })
                    }else{
                        console.log(res.data.message)
                        this.openLoading=false // stop loading
                        Toast.fail('Login failed!')
                    }
                })
                .catch((error) => {
                    console.log(error)
                    Toast.fail('Login failed,error!')
                    this.openLoading=false  // stop loading 
                })
            }
        },
    }
</script>

<style lang="scss" scoped>
    .register-panel{
        width:96%;
        border-radius: 5px;
        margin:20px auto;
        padding-bottom:50px;
        .register-button{
            padding-top:10px;
        }
    }
</style>