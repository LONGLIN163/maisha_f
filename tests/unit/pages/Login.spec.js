
import { mount, createLocalVue} from '@vue/test-utils'
import Login from '@/components/pages/Login.vue'


import VueRouter from 'vue-router';
const Vant=require("vant")

const localVue = createLocalVue();
localVue.use(VueRouter); 
localVue.use(Vant); 

const router = new VueRouter();

describe('Login',  () => {

  let wrapper;

  // mock loginAction()
  const loginAction = jest.spyOn(Login.methods, 'loginAction')
    .mockImplementation(() => {
      //console.log("1. mock loginAction implementation")
      checkForm() && login() // call checkForm(set true), then call login
  })
  // mock checkForm()
  const checkForm = jest.spyOn(Login.methods, 'checkForm')
    .mockImplementation(() => {
      //console.log("2. mock checkForm implementation")
      return true
  })
  // mock login()
  const login = jest.spyOn(Login.methods, 'login')
    .mockImplementation(() => {
      //console.log("3. mock login implementation")
  })

  beforeEach(() => {
    wrapper = mount(Login,{
      localVue,
      router
    })
  })


  it('renders Login',async () => {
    expect(wrapper).toMatchSnapshot()
  })

  // ***test goback in e2e test***

  it('test login input field ',async () => {
    await wrapper.find('[data-test="username"]').setValue("usr");
    expect(wrapper.find('[data-test="username"]').element.value).toBe('usr')

    await wrapper.find('[data-test="password"]').setValue("pwd");
    expect(wrapper.find('[data-test="password"]').element.value).toBe('pwd')
  })

  it('loginAction() should have been called correctly ',async () => {
    await wrapper.find('[data-test="loginsubmit"]').trigger('click');
    expect(loginAction).toHaveBeenCalled()
  })

})