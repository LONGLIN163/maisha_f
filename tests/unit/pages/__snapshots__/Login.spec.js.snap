// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders Login 1`] = `
<div>
  <div class="van-nav-bar van-hairline--bottom" data-test="logintitle">
    <div class="van-nav-bar__content">
      <div class="van-nav-bar__left"><i class="van-icon van-icon-arrow-left van-nav-bar__arrow">
          <!---->
        </i><span class="van-nav-bar__text">back</span></div>
      <div class="van-nav-bar__title van-ellipsis">Login</div>
    </div>
  </div>
  <div class="register-panel">
    <div class="van-cell van-cell--required van-field">
      <div class="van-cell__title van-field__label"><span>User Name</span></div>
      <div class="van-cell__value van-field__value">
        <div class="van-field__body"><input type="text" data-test="username" placeholder="Please input user name" class="van-field__control"></div>
      </div>
    </div>
    <div class="van-cell van-cell--required van-field">
      <div class="van-cell__title van-field__label"><span>password</span></div>
      <div class="van-cell__value van-field__value">
        <div class="van-field__body"><input type="password" data-test="password" placeholder="Please input password" class="van-field__control"></div>
      </div>
    </div>
    <div class="register-button"><button data-test="loginsubmit" class="van-button van-button--primary van-button--large">
        <div class="van-button__content"><span class="van-button__text">Login</span></div>
      </button></div>
  </div>
</div>
`;
